pandas/_libs/index.pyx

```diff
@@ -275,6 +275,14 @@ cdef class IndexEngine:
 
         self._ensure_mapping_populated()
         values = np.array(self._get_index_values(), copy=False)
+        targets_mask = isnaobj(targets)
```

Comment:
any particular reason you're doing this up-front instead of in the loop below?

Reply:
I did this up-front because when you store NaN in a stargets (line 286), the `val in stargets` (line 318) statement does not work